src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,668 @@
+        get_contour_pt_stub = tmp9;
+
+       /* Having now created the font upcall stubs, we can call down to create
+        * the native harfbuzz object holding these.
+        */
+        FunctionDescriptor createFontFuncsDescriptor = FunctionDescriptor.of(
+            ADDRESS,     // hb_font_funcs* return type
+            ADDRESS,     // glyph_fn upcall stub
+            ADDRESS,     // variation_fn upcall stub
+            ADDRESS,     // h_advance_fn upcall stub
+            ADDRESS,     // v_advance_fn upcall stub
+            ADDRESS);     // contour_pt_fn upcall stub
+        Optional<MemorySegment> create_font_funcs_symbol = SYM_LOOKUP.find("HBCreateFontFuncs");
+        @SuppressWarnings("restricted")
+        MethodHandle create_font_funcs_handle =
+            LINKER.downcallHandle(create_font_funcs_symbol.get(), createFontFuncsDescriptor);
+
+        MemorySegment s = null;
+        try {
+            s = (MemorySegment)create_font_funcs_handle.invokeExact(
+                get_nominal_glyph_stub,
+                get_var_glyph_stub,
+                get_h_advance_stub,
+                get_v_advance_stub,
+                get_contour_pt_stub);
+       } catch (Throwable t) {
+          t.printStackTrace();
+       }
+       hb_jdk_font_funcs_struct = s;
+
+       FunctionDescriptor store_layout_fd =
+          FunctionDescriptor.ofVoid(
+                   JAVA_INT,               // slot
+                   JAVA_INT,               // baseIndex
+                   JAVA_INT,               // offset
+                   JAVA_FLOAT,             // startX
+                   JAVA_FLOAT,             // startX
+                   JAVA_FLOAT,             // devScale
+                   JAVA_INT,               // charCount
+                   JAVA_INT,               // glyphCount
+                   ADDRESS,                // glyphInfo
+                   ADDRESS);               // glyphPos
+       MethodHandle store_layout_mh =
+          getMethodHandle("store_layout_results", store_layout_fd);
+       @SuppressWarnings("restricted")
+       MemorySegment tmp10 = LINKER.upcallStub(store_layout_mh, store_layout_fd, garena);
+       store_layout_results_stub = tmp10;
+
+        long maxpos = Long.MAX_VALUE / PositionLayout.byteSize();
+        SequenceLayout glyphPosLayout = MemoryLayout.sequenceLayout(maxpos, PositionLayout);
+        x_offsetHandle = getVarHandle(glyphPosLayout, "x_offset");
+        y_offsetHandle = getVarHandle(glyphPosLayout, "y_offset");
+        x_advanceHandle = getVarHandle(glyphPosLayout, "x_advance");
+        y_advanceHandle = getVarHandle(glyphPosLayout, "y_advance");
+        long maxinfo = Long.MAX_VALUE / GlyphInfoLayout.byteSize();
+        SequenceLayout glyphInfosLayout = MemoryLayout.sequenceLayout(maxinfo, GlyphInfoLayout);
+        codePointHandle = getVarHandle(glyphInfosLayout, "codepoint");
+        clusterHandle = getVarHandle(glyphInfosLayout, "cluster");
```

Comment:
There are better ways to do this in the latest API, by using the extra offset parameter and `MemoryLayout::scaleHandle`.

I suggest changing this to:

```suggestion
        x_offsetHandle = getVarHandle(PositionLayout, "x_offset");
        y_offsetHandle = getVarHandle(PositionLayout, "y_offset");
        x_advanceHandle = getVarHandle(PositionLayout, "x_advance");
        y_advanceHandle = getVarHandle(PositionLayout, "y_advance");
        codePointHandle = getVarHandle(GlyphInfoLayout, "codepoint");
        clusterHandle = getVarHandle(GlyphInfoLayout, "cluster");
```

And then in `getVarHandle`:

```Java
    private static VarHandle getVarHandle(MemoryLayout enclosing, String name) {
        VarHandle h = layout.varHandle(PathElement.groupElement(name));
        // scale offset by the size of 'enclosing'
        h = MethodHandles.collectCoordinates(h, 1, enclosing.scaleHandle());
        /* insert 0 offset so don't need to pass arg every time */
        return MethodHandles.insertCoordinates(h, 1, 0L).withInvokeExactBehavior();
    }
```

Reply:
I've just pushed this change, but I'm unclear why it is "better". It seems more obscure to me.
I've added a somewhat expanded comment of what I guess this does to help
me and others understand what it does. Please take a look to see if it makes sense.

Reply:
>  I'm unclear why it is "better". It seems more obscure to me.

Ok. I think it's better because it doesn't require creating a maximum size sequence layout in order to then make a var handle out of, which is a bit of a hack IMO. One that was required in the previous version of the API.

This kind of use-case, where the size of the sequence is not known statically, is one of the reasons why we added the extra base offset parameter to the var handles.

Another way of writing this would be to use the base var handle, with its extra leading offset parameter, and then pass e.g. `i * PositionLayout.byteSize()` as the offset at every call site. The two extra combination steps essentially create  var handle with that behavior baked in.

Reply:
> > I'm unclear why it is "better". It seems more obscure to me.
> 
> Ok. I think it's better because it doesn't require creating a maximum size sequence layout in order to then make a var handle out of, which is a bit of a hack IMO. One that was required in the previous version of the API.

Not sure which "previous" that was, but in JDK 21 I did not need to specify a size.
The need to do that was something that came in as of JDK 22 and I thought it a bit of a backwards step perhaps motivated to help devs understand the sizes involved but given the arithmetic involved in general I am not sure it was justified.

> 
> This kind of use-case, where the size of the sequence is not known statically, is one of the reasons why we added the extra base offset parameter to the var handles.

The previous API dealt with that just fine. And equivalently as far as I can  tell.
The base offset parameter may have other uses but I need  its relevance to this explained to me,

> Another way of writing this would be to use the base var handle, with its extra leading offset parameter, and then pass e.g. `i * PositionLayout.byteSize()` as the offset at every call site (where `i` is the array index). The two extra combination steps essentially create a var handle with that behavior baked in.

OK, but now you have a VarHandle intended for use on a SequenceLayout (ie array) of Struct and are disguising it for some reason that isn't (sufficiently?) obvious at an API level and definitely isn't obvious at a performance level.
Is there some fundamental reason why the 21 API could not internally be reduced to the same ?

Reply:
Requiring users to specify the size of the sequence layout was done in order to dispel the illusion that there was any kind of special handling for a sequence layout created without a size. e.g. if you try to allocate with it, what should happen? Should we detect that this as a special case? Or just crush with an OOME? This is something other users ran into in practice, and removing the size-less factory revealed some latent bugs in the tests as well. So, I feel that overall, dropping the size-less factory was the right move. This was more or less an orthogonal decision to the decision of adding the base offset parameter.

The previous JDK 21 API asked users to construct layouts for memory of which they did not know the layout in advance. e.g. when creating a var handle from a sequence layout with the maximum number of elements, the in-memory array that is being accessed is likely not actually an array with the maximum number of elements? The special max element sequence layout is just a workaround used to be able to create an indexed var handle.

Another example is a 2-dimensional matrix with an dynamic row and column size. How should this be represented using a memory layout? We can't use the max element sequence layout trick in that case, since the size of the inner sequence affects the scaling of the index for the outer sequence.

The core issue is that, to get good performance, a user needs to construct the layout and derive var handles in advance, but at the same time it is not possible to represent a layout with a 'dynamic' size. We went back and forth on ideas in order to add better support for dynamic sizes in the layout API. But in the end, all the things we tried ended up being convoluted, and had their own set of corner cases that were ill-addressed.

So, the conclusion we arrived at was that layouts are better left alone, and should only be used to represent memory layouts that are 'static'/fixed and known up front. In that case a user can declare the layout, and all the var handles they need, in advance, and stick them into `static final` fields, which is required to get good performance.

But then the question becomes: what about structural access to memory whose layout can _not_ be represented statically? Even in those cases, often there is a part of the structure of the memory layout that is fixed, and part of the layout that is dynamic. The memory layout API can still be used for the fixed part, and then the extra base offset parameter can be used to handle the dynamic part, by injecting external/ad-hoc offset computations into the var handle access.

As far as I can tell in this patch, the var handles that are created are for arrays whose size is not known statically, so this would fall into the territory that is address by the extra base offset parameter. That's why I think it is 'better' to use that mechanism. It avoids the need for the max element sequence layout workaround, which, as you found, also became more cumbersome due to the decision to drop the size-less sequence layout factory. At the end of the day, it was just a suggestion. I think you should stick with the version you prefer. (technically the version using the sequence layout adds an extra bounds check, but I think it will be folded away by C2).

If the var handle combinator steps look obscure, I think that is a good indicator that we perhaps need an extra `arrayVarHandle` convenience method that does the combination steps.

Reply:
I feel like I'm starting to learn that once you've got your VarHandle from FFM, you need to head over to java.lang.invoke and hunt around to see what tricks you can do with it.
The less I have to do that the better.

Reply:
I wouldn't say it's necessarily required, but I would say that it is what ultimately leads to the most succinct code. I'm probably biased towards using the java.lang.invoke tricks though.

The java.lang.invoke combinator API is a convenient way to generate wrapper functions around a VarHandle (or MethodHandle). But, you could also write out the wrapper functions manually. In this case that shouldn't be too much of a problem given that the code only uses `get`. You could declare basic field VarHandles created with `layout.varHandle(PathElement.groupElement(name));`, and then write the wrapper functions like so:

```
public static int getYOffset(MemorySegment glyphPosArr, int index) {
    long posArrOffset = PositionLayout.scale(0, index); // or just: index * PositionLayout.byteSize()
    return (int) y_offsetHandle.get(glyphPosArr, posArrOffset);
}
```

You would need 6 of these in total. Or, just use the implementation 'inline', without the wrapper functions.